src/self-managed-debug/src/k8s_resource_dumper.rs

```diff
@@ -0,0 +1,303 @@
+            let file_name = file_path.join(format!("{}.{}.log", pod_name, suffix));
+
+            let logs = pods
+                .logs(
+                    pod_name,
+                    &LogParams {
+                        previous: is_previous,
+                        timestamps: true,
+                        ..Default::default()
+                    },
+                )
+                .await?;
+
+            if logs.is_empty() {
+                eprintln!("No {} logs found for pod {}", suffix, pod_name);
+                return Ok(());
+            }
+
+            let mut file = File::create(&file_name)?;
+            file.write_all(logs.as_bytes())?;
+            println!("Exported {}", file_name.display());
+
+            Ok(())
+        }
+
+        if let Err(e) = export_pod_logs(&pods, &pod_name, &file_path, true).await {
+            let print_error = || {
+                eprintln!(
+                    "Failed to export previous logs for pod {}: {}",
+                    &pod_name, e
+                );
+            };
+
+            if let Some(kube::Error::Api(e)) = e.downcast_ref::<kube::Error>() {
+                if e.code == 400 {
+                    eprintln!("No previous logs available for pod {}", pod_name);
+                } else {
+                    print_error();
+                }
+            } else {
+                print_error();
+            }
+        }
+
+        if let Err(e) = export_pod_logs(&pods, &pod_name, &file_path, false).await {
+            eprintln!("Failed to export current logs for pod {}: {}", &pod_name, e);
+        }
+    }
+    Ok(())
+}
+
+/// Write k8s pod logs to a yaml file per pod.
+async fn dump_k8s_pod_logs(context: &Context, client: Client, namespace: &String) -> () {
+    if let Err(e) = _dump_k8s_pod_logs(context, client, namespace).await {
+        eprintln!("Failed to dump k8s pod logs: {}", e);
+    }
+}
+
+/// Write namespace-level k8s resources to a yaml file per resource.
+pub async fn dump_namespaced_resources(context: &Context, client: &Client, namespace: String) {
+    K8sResourceDumper::<Pod>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
+    K8sResourceDumper::<Service>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
+    K8sResourceDumper::<Deployment>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
+    K8sResourceDumper::<StatefulSet>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
+    K8sResourceDumper::<ReplicaSet>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
+    K8sResourceDumper::<NetworkPolicy>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
+    K8sResourceDumper::<Event>::namespaced(context, client.clone(), namespace.clone())
+        .dump()
+        .await;
```

Comment:
Looking at your list of api-resources, there are a few things that might make sense to include:

High priority:
- materializes
- roles
- rolebindings
- configmaps
- secrets (probably only the names, maybe with an arg saying whether to include the contents)

Lower priority:
- daemonsets (may want to get this from all namespaces, since they can affect the node even if they are in a separate namespace)
- storageclasses
- persistentvolumes
- persistentvolumeclaims
- serviceaccounts
- mutatingwebhookconfigurations
- validatingwebhookconfigurations
- customresourcedefinitions
- certificates (I mean the ones from cert-manager, you probably need the full name, and they may not exist)

Reply:
> materializes

Ah what are materializes?  Everything else I can try to get!

Reply:
Was able to get everything but:

> certificates

> daemonsets (may want to get this from all namespaces, since they can affect the node even if they are in a separate namespace)

Not too sure where the certificates in cert-manager live but if they're in :secrets, then seems like getting the secret names cover it! For daemonsets, given their a namespaced resource, I need to actually define a namespaced resource. I'm also trying to reduce the number of `:all` API calls given I don't want the query to fail in case the user doesn't have RBAC permissions for some of the namespaces.